From ee5a16c8a402d5c274317d239b22dbe9ed2566d1 Mon Sep 17 00:00:00 2001
From: Francesco Valla <valla.francesco@gmail.com>
Date: Sun, 17 Jul 2016 23:27:08 +0200
Subject: [PATCH 1/3] fs: jffs2: Add setup code for spi nor flash.

Flash need to be setup, while using it for Jffs2 filesystem.
The setup code checks was not present for serial flash like
m25p80. As a result of which, internal pointers/buffers are in unknown
state. This leads to
a. dumps like below[1] while mounting a jffs2 filesystem.
b. Hang while doing a mount second time.

[1]:
root at am437x-evm:~# sync
[   84.112640] INFO: trying to register non-static key.
[   84.117553] the code is fine but needs lockdep annotation.
[   84.122985] turning off the locking correctness validator.
[   84.128417] CPU: 0 PID: 1289 Comm: sync Not tainted 3.12.0-67367-g2d1978c-dirty #18
[   84.136016] [<c001bcb8>] (unwind_backtrace+0x0/0xf0) from [<c0017dd0>] (show_stack+0x10/0x14)
[   84.144439] [<c0017dd0>] (show_stack+0x10/0x14) from [<c05a69c8>] (dump_stack+0x78/0x94)
[   84.152465] [<c05a69c8>] (dump_stack+0x78/0x94) from [<c009e2e0>] (__lock_acquire+0x1788/0x1b8c)
[   84.161132] [<c009e2e0>] (__lock_acquire+0x1788/0x1b8c) from [<c009ec30>] (lock_acquire+0x9c/0x104)
[   84.170104] [<c009ec30>] (lock_acquire+0x9c/0x104) from [<c005ede8>] (flush_work+0x38/0x78)
[   84.178375] [<c005ede8>] (flush_work+0x38/0x78) from [<c005eef0>] (__cancel_work_timer+0x84/0x124)
[   84.187225] [<c005eef0>] (__cancel_work_timer+0x84/0x124) from [<c024c1d4>] (jffs2_sync_fs+0x14/0x38)
[   84.196380] [<c024c1d4>] (jffs2_sync_fs+0x14/0x38) from [<c013c138>] (sync_fs_one_sb+0x28/0x2c)
[   84.204986] [<c013c138>] (sync_fs_one_sb+0x28/0x2c) from [<c0115360>] (iterate_supers+0xb0/0xd8)
[   84.213684] [<c0115360>] (iterate_supers+0xb0/0xd8) from [<c013c228>] (sys_sync+0x3c/0x98)
[   84.221862] [<c013c228>] (sys_sync+0x3c/0x98) from [<c0013dc0>] (ret_fast_syscall+0x0/0x48)

The patch fixes the above two issue.

Signed-off-by: Sourav Poddar <sourav.poddar@ti.com>
Signed-off-by: George Cherian <george.cherian@ti.com>
Signed-off-by: Francesco Valla <valla.francesco@gmail.com>
---
 fs/jffs2/fs.c       | 7 +++++++
 fs/jffs2/os-linux.h | 2 ++
 2 files changed, 9 insertions(+)

diff --git a/fs/jffs2/fs.c b/fs/jffs2/fs.c
index 2caf168..da934cb 100644
--- a/fs/jffs2/fs.c
+++ b/fs/jffs2/fs.c
@@ -731,6 +731,13 @@ static int jffs2_flash_setup(struct jffs2_sb_info *c) {
 			return ret;
 	}
 
+	/* and m25p80 spi flash */
+	if (jffs2_nor_spi_wbuf_flash(c)) {
+		ret = jffs2_nor_wbuf_flash_setup(c);
+		if (ret)
+			return ret;
+	}
+
 	/* and an UBI volume */
 	if (jffs2_ubivol(c)) {
 		ret = jffs2_ubivol_setup(c);
diff --git a/fs/jffs2/os-linux.h b/fs/jffs2/os-linux.h
index 824e61e..e0112d4 100644
--- a/fs/jffs2/os-linux.h
+++ b/fs/jffs2/os-linux.h
@@ -90,6 +90,7 @@ static inline void jffs2_init_inode_info(struct jffs2_inode_info *f)
 #define jffs2_dataflash_setup(c) (0)
 #define jffs2_dataflash_cleanup(c) do {} while (0)
 #define jffs2_nor_wbuf_flash(c) (0)
+#define jffs2_nor_spi_wbuf_flash(c) (0)
 #define jffs2_nor_wbuf_flash_setup(c) (0)
 #define jffs2_nor_wbuf_flash_cleanup(c) do {} while (0)
 #define jffs2_ubivol(c) (0)
@@ -134,6 +135,7 @@ int jffs2_ubivol_setup(struct jffs2_sb_info *c);
 void jffs2_ubivol_cleanup(struct jffs2_sb_info *c);
 
 #define jffs2_nor_wbuf_flash(c) (c->mtd->type == MTD_NORFLASH && ! (c->mtd->flags & MTD_BIT_WRITEABLE))
+#define jffs2_nor_spi_wbuf_flash(c) (c->mtd->type == MTD_NORFLASH && (c->mtd->flags & MTD_CAP_NORFLASH))
 int jffs2_nor_wbuf_flash_setup(struct jffs2_sb_info *c);
 void jffs2_nor_wbuf_flash_cleanup(struct jffs2_sb_info *c);
 void jffs2_dirty_trigger(struct jffs2_sb_info *c);
-- 
2.7.4

